src/monitoring_tests/combinatorial_auction_surplus_test.py

```diff
@@ -0,0 +1,266 @@
+                f"Tx Hash: {competition_data['transactionHash']}",
+                f"Winning Solver: {solution['solver']}",
+                f"Winning surplus: {aggregate_solution}",
+                f"Baseline surplus: {baseline_surplus}",
+                f"Solutions filtering winner: {filter_mask[-1]}",
+                f"Solvers filtering winner: {solutions_filtering_winner}",
+                f"Combinatorial winners: {winning_solvers}",
+                f"Total surplus: {float(total_surplus):.5f} ETH",
+                f"Combinatorial surplus: {float(total_combinatorial_surplus):.5f} ETH",
+                f"Absolute difference: {float(a_abs_eth):.5f}ETH",
+            ]
+        )
+        if a_abs_eth > SURPLUS_ABSOLUTE_DEVIATION_ETH:
+            self.alert(log_output)
+        elif (
+            a_abs_eth > SURPLUS_ABSOLUTE_DEVIATION_ETH / 10
+            or not len(filter_mask[-1]) == 0
+        ):
+            self.logger.info(log_output)
+        else:
+            self.logger.debug(log_output)
+
+        return True
+
+    def get_uid_trades(self, solution: dict[str, Any]) -> dict[str, Trade]:
+        """Get a dictionary mapping UIDs to trades in a solution."""
+        calldata = solution["callData"]
+        settlement = self.web3_api.get_settlement_from_calldata(calldata)
+        trades = self.web3_api.get_trades(settlement)
+        trades_dict = {
+            solution["orders"][i]["id"]: trade for (i, trade) in enumerate(trades)
+        }
+        return trades_dict
+
+    def get_token_pairs_surplus(
+        self, solution: dict[str, Any], prices: dict[str, float]
+    ) -> dict[tuple[str, str], Fraction]:
+        """Aggregate surplus of a solution on the different token pairs.
+        The result is a dict containing directed token pairs and the aggregated surplus on them.
+
+        Instead of surplus we use the minimum of surplus and the objective. This is more
+        conservative than just using objective. If fees are larger than costs, the objective is
+        larger than surplus and surplus is used for the comparison. If fees are larger than costs,
+        the objective is smaller than surplus and the objective is used instead of surplus for
+        filtering. This takes care of the case of solvers providing a lot of surplus but at really
+        large costs.
+        """
+        trades_dict = self.get_uid_trades(solution)
+        surplus_dict: dict[tuple[str, str], Fraction] = {}
+        for uid in trades_dict:
+            trade = trades_dict[uid]
+            token_pair = (trade.data.sell_token.lower(), trade.data.buy_token.lower())
+
+            # compute the conversion rate of the surplus token to ETH
+            if trade.get_surplus_token() == trade.data.sell_token:
+                surplus_token_to_eth = Fraction(
+                    int(prices[token_pair[0]]),
+                    10**36,
+                )
+            elif trade.get_surplus_token() == trade.data.buy_token:
```

Comment:
is it possible there is a third case?

Reply:
No. Should I rewrite as `else` instead of `elif`? Or add some error message in case both checks fail?

Reply:
I rewrote this as `if "is sell order": ... else:`